text/0580-destroyables.md

```diff
@@ -0,0 +1,385 @@
+
+class Modal extends Component {
+  @service resize;
+
+  constructor() {
+    this.resize.register(this, this.layout);
+
+    registerDestructor(this, () => this.resize.unregister(this));
+  }
+}
+```
+
+Multiple destructors can be associated with a given destroyable, and they can be
+associated over time, allowing libraries like `ember-lifeline` to dynamically
+add destructors as needed. `registerDestructor` also returns the associated
+destructor function, for convenience.
+
+The destructor function is passed a single argument, which is the destroyable
+itself. This allows the function to be reused multiple times for many
+destroyables, rather than creating a closure function per destroyable.
+
+```js
+import { registerDestructor } from '@ember/destroyable';
+
+function unregisterResize(instance) {
+  instance.resize.unregister(instance);
+}
+
+class Modal extends Component {
+  @service resize;
+
+  constructor() {
+    this.resize.register(this, this.layout);
+
+    registerDestructor(this, unregisterResize);
+  }
+}
+```
+
+- Registering a destructor on a destroyed object should throw an error.
+- Attempting to register the same destructor multiple times should throw an
+  error.
+
+#### `unregisterDestructor`
+
+Receives a destroyable and a destructor function, and de-associates the
+destructor from the destroyable.
+
+```js
+import { unregisterDestructor } from '@ember/destroyable';
+
+class Modal extends Component {
+  @service modals;
+
+  constructor() {
+    this.modals.add(this);
+
+    this.modalDestructor = registerDestructor(this, () => this.modals.remove(this));
+  }
+
+  @action pinModal() {
+    unregisterDestructor(this, this.modalDestructor);
+  }
+}
+```
+
+- Calling `unregisterDestructor` on a destroyed object should throw an error.
+- Calling `unregisterDestructor` with a destructor that is not associated with
+  the object should throw an error.
+
+#### `destroy`
+
+`destroy` initiates the destruction of a destroyable object. It runs all
+associated destructors, and then destroys all children recursively.
+
+```js
+let obj = {};
+
+registerDestructor(obj, () => console.log('destroyed!'));
+
+destroy(obj); // this will schedule the destructor to be called
+
+// ...some time later, during scheduled destruction
+
+// destroyed!
+```
+
+Destruction via `destroy()` follows these steps:
+
+1. Mark the destroyable such that `isDestroying(destroyable)` returns `true`
+2. Schedule calling the destroyable's destructors
+3. Call `destroy()` on each of the destroyable's associated children
+4. Schedule setting destroyable such that `isDestroyed(destroyable)` returns `true`
+
+This algorithm results in the entire tree of destroyables being first marked as
+destroying, then having all of their destructors called, and finally all being
+marked as `isDestroyed`. There won't be any in between states where some items
+are marked as `isDestroying` while destroying, while others are not.
```

Comment:
I overlooked the keyword `Schedule`, meaning that the action happens asynchronously / next tick / etc. and thought that this algorithm actually results in marking the parent, calling the parent destructors already and only _then_ marking the children and calling the children's destructors.

Just adding this here, in case I am not the only one.

Do we wanna make any guarantees with regards to the timing of scheduling? Is it scheduled onto a specific runloop queue, next tick, or no guarantees on that at all?

Reply:
We do not want to make guarantees about timing, as timing could change per-environment, or could change in the future. The only guarantee is relative timing, as outlined in this RFC.